core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -468,12 +469,21 @@ public int compareTo(Key o) {
 
   @Immutable
   private static class MergedResult implements Result {
-    private final Optional<TransactionResult> result;
     private final Put put;
+    private final Map<String, Value<?>> values;
 
     public MergedResult(Optional<TransactionResult> result, Put put) {
-      this.result = result;
       this.put = put;
+
+      Map<String, Value<?>> tmp = new HashMap<>();
+      if (result.isPresent()) {
+        tmp.putAll(result.get().getValues());
+      } else {
+        put.getPartitionKey().get().forEach(v -> tmp.put(v.getName(), v));
+        put.getClusteringKey().ifPresent(c -> c.get().forEach(v -> tmp.put(v.getName(), v)));
```

Comment:
This logic includes the primary key into `values` when the result in the read set is empty.